sktree/tree/_unsup_criterion.pyx

```diff
@@ -358,42 +492,99 @@ cdef class TwoMeans(UnsupervisedCriterion):
+        cdef double mean_left
+        cdef double mean_right
+        cdef double sig_left
+        cdef double sig_right
+        cdef double BIC_diff_var_left
+        cdef double BIC_diff_var_right
+        cdef double BIC_same_var_left
+        cdef double BIC_same_var_right
+
+        # number of samples of left and right
+        s_l = pos - start
+        s_r = end - pos
+
+        # compute prior (i.e. \hat{\pi_1} and \hat{\pi_2} in the paper)
+        p_l = s_l / self.n_node_samples
+        p_r = s_r / self.n_node_samples
 
         # first compute mean of left and right
         mean_left = self.sum_left / self.weighted_n_left
         mean_right = self.sum_right / self.weighted_n_right
 
-        # set values at the address pointer is pointing to with the variance
-        # of the left and right child
-        impurity_left[0] = self.sum_of_squares(
+        sig_left = self.sum_of_squares(
             start,
             pos,
             mean_left
-        ) / self.weighted_n_left
-        impurity_right[0] = self.sum_of_squares(
+        )
+
+        sig_right = self.sum_of_squares(
             pos,
             end,
             mean_right
-        ) / self.weighted_n_right
+        )
 
-    cdef void set_sample_pointers(
-        self,
-        SIZE_t start,
-        SIZE_t end
-    ) nogil:
-        """Set sample pointers in the criterion.
+        sig_comb = (sig_left + sig_right) / (self.weighted_n_left + self.weighted_n_right)
 
-        Set given start and end sample_indices. Also will update node statistics,
-        such as the `sum_total`, which tracks the total value within the current
-        node for sample_indices[start:end].
+        sig_left = sig_left / self.weighted_n_left
+        sig_right = sig_right / self.weighted_n_right
 
-        Parameters
-        ----------
-        start : SIZE_t
-            The start sample pointer.
-        end : SIZE_t
-            The end sample pointer.
-        """
-        self.n_node_samples = end - start
-        self.start = start
-        self.end = end
+        # BIC score computed using left and right variances
+        # -2(n_1\log{\hat{w}_1}-\frac{n_1}{2}\log{2\pi\hat{\sigma}_{1}^2} - n_2\log{\hat{w}_2}+\frac{n_2}{2}\log{2\pi\hat{\sigma}_{2}^2})
+        BIC_diff_var_left = -2*(s_l*(log(p_l) - log(2*pi*sig_left)/2))
+        BIC_diff_var_right = -2*(s_r*(log(p_r) - log(2*pi*sig_right)/2))
+
+        # BIC score computed using combined variances
+        # -2(n_1\log{\hat{w}_1}-\frac{n_1}{2}\log{2\pi\hat{\sigma}_{comb}^2} - n_2\log{\hat{w}_2}+\frac{n_2}{2}\log{2\pi\hat{\sigma}_{comb}^2})
+        BIC_same_var_left = -2*(s_l*(log(p_l) - log(2*pi*sig_comb)/2))
+        BIC_same_var_right = -2*(s_r*(log(p_r) - log(2*pi*sig_comb)/2))
+
+        # simplified equation of maximum log likelihood function 
+        # at corresponding sample size for left and right child
+        if BIC_diff_var_left - BIC_diff_var_right < BIC_same_var_left - BIC_same_var_right:
```

Comment:
You'll need to double check the negative signs all make sense. 

The implementation of splitter and criterion try to create splits that minimizes some criterion metric (Gini, variance, etc.). 

You'll have to make sure your implementation lines up with that.

Reply:
I will look into it to understand the context. FYI, I tried 1. negative sign 2. switching greater than to less than 3. comparing them to absolute values

Reply:
Oh I mean the children impurities might need to have a negative sign

Reply:
I tried that as well. 1. Negative both 2. Negative left only 3. Negative right only 4. Reversing signs within the two terms that make up left and right children.